### PR TITLE
Migrate netcdf-cxx4 to use CMake and stop linking against system libnetcdf

### PR DESCRIPTION
**Pull request by Chrismarsh**
This PR fixes issue #42735 

The underlying issue is that some of the assumptions made in netcdf-cxx4 package.py result in system libnetcdf being linked to.

This PR migrates to using CMake to better control how dependencies are found, explicitly depends upon netcdf and hdf5 as netcdf-cxx4 uses symbols from both, and cleans up the logic to be more composable and consistent in how MPI etc are used.

Specifically:
- `hdf5` is searched from in the netcdf-cxx4 CMakeLists.txt, so hdf5 has been added as an explicit dependency
- The logic around `^hdf5+mpi` has been cleaned up and now the spack MPI compiler wrapper is used.  The CC flag mutation for including MPI is not composable and should be done using external package definitions for MPI.
- The patch removing `libh5bzip2` is removed as this library is (seemingly) not installed via cmake
- VERSION is not written out by cmake so this patch is removed
- An option to enable tests was added, but this looks to have a lot of internal unidata config, so I am not sure how useful it is.

### CODE REVIEW

**Comment by Chrismarsh** [comment]
@spackbot fix style

**Comment by Chrismarsh** [comment]
Tagging @climbfuji @scheibelp @skosukhin from pull #29246 as they had eyes on this last. I want to make sure these changes continue to work for them.

I've tested this on macos and rehl with an external intel-mpi

**Comment by skosukhin** [comment]
I would prefer having the `CMake` support as an alternative to `Autotools` like in [`netcdf-c`](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/netcdf-c/package.py) (with the latter by default).

As a side note, I would recommend any developer who considers using `netcdf-cxx` in their project use `netcdf-c` instead.

**Comment by Chrismarsh** [comment]
In your opinion, what is the benefit to maintaining the two code paths? Especially given the autotools variant was much more complex and seemingly error prone?

**Comment by skosukhin** [comment]
@Chrismarsh mainly because I went through a similar transition with the `hdf5` package and I did not like the experience (it still hurts after 3 years). I also disagree with "Especially given the autotools variant was much more complex and seemingly error prone". The current implementation might seem complex but not without reasons and I don't see (maybe yet) how those reasons are irrelevant to the `CMake` implementation.

**Comment by Chrismarsh** [comment]
> I went through a similar transition with the hdf5 package and I did not like the experience

Specifically with migrating `hdf5/package.py` to cmake? or something else?

> I don't see (maybe yet) 

The main issue for me is that [the way `libnetcdf` was linked](https://github.com/spack/spack/blob/f9e9fab2daa1f43b49f58425f50b6a7e66bd137b/var/spack/repos/builtin/packages/netcdf-cxx4/package.py#L45) in [doesn't work on systems with more than 1 netcdf-c](https://github.com/spack/spack/issues/42735).  There is not sufficient fine grained control over the paths.  As a result I ended up with a netcdf-cxx4 linked against the wrong netcdf-c, which is obviously non-optimal. 

In an effort to try to figure out what was going on, I quickly converted the package.py to a cmake variant. I am more experienced with cmake than autotools, and cmake provides (to me) more capabiilities to figure out what was going on.

As a result, the linking is correctly handled because the `find_package` of the netcdf-cxx4 CMakeLists.txt handles it automatically. Then once I started to look through the rest of the package.py, it became obvious that all the extra complexity goes away when using the cmake approach. This includes removing the non-composable way mpicxx is used via direct manipulation of `CC=`.

My perspective is:
- Autotools package.py doesn't currently work as expected
- It works as expected with CMake
- All the extra complexity of the Autotools build goes away when using cmake.

That is, it wasn't clear to me the added complexity of the autotools build was warranted, and thus I removed it for this PR.

It is obiously trivial to include the existing Autotools building, and if that is what is needed to get this PR merged, then I can admend. However, I cannot help maintain the autotools part and the code duplicity is not immediately obvious to me unless there are workflows that literally require autotools?

File: var/spack/repos/builtin/packages/netcdf-cxx4/package.py
Comment (scheibelp) [inline var/spack/repos/builtin/packages/netcdf-cxx4/package.py]: (minor) As originally implemented, this would have allowed for `spec[netdf-cxx4:shared].libs`, although I don't see any packages making use of that, I also don't think that this is necessary for the other changes. Did you encounter an issue with this while you were making other changes?

**Review by scheibelp — COMMENTED** [review]
Overall this looks good to me. I have a minor request. I want to make sure we address concerns before merging though:

* It looks like this supports all variants that were supported before
* I was initially worried that older versions might not support CMake builds, but I only see two versions here and they are very "close" (4.3.0 and 4.3.1)

I also get the impression that CMake would in fact handle some of the original logic that was needed for Autotools: e.g.

```
flags.append(self.spec["netcdf-c"].libs.link_flags)
```

Our CI currently does not cover `netcdf-cxx4`, but it sounds like the PR author also built with `^hdf5+mpi`, which ought to cover concerns related to removal of `CC=...`.

**Comment by scheibelp** [comment]
Some of these changes undo work in https://github.com/spack/spack/pull/29246 which fixes a build error for Mac OS on https://github.com/spack/spack/issues/26559. I could try building with this PR on Mac OS to check that. @Chrismarsh It would be good to know what system/variations you built.

**Comment by climbfuji** [comment]
@Chrismarsh Thanks for tagging me. I also want to make sure this works on macOS. Give me a few days and I'll test it out.

**Comment by Chrismarsh** [comment]
Sorry, I should have stated that more clearly. Yes, I have tested this on Macos (13.3):
```
spack install netcdf-cxx4 ^hdf~mpi
spack install netcdf-cxx4 ^hdf5+mpi
```
I don't experience the issue #26559, but it would be good to triple check on other machines.

I have also tested on REHL against oneapi-mpi

File: var/spack/repos/builtin/packages/netcdf-cxx4/package.py
Comment (skosukhin) [reply var/spack/repos/builtin/packages/netcdf-cxx4/package.py]: (major) The original implementation also returns `None` if the libraries are not found. It's worse than throwing [an exception](https://github.com/spack/spack/blob/2d8e0825feb05f082f5dc4b4c3832fc323f7cf07/var/spack/repos/builtin/packages/netcdf-fortran/package.py#L111-L114) but much better than an empty list.

**Comment by skosukhin** [comment]
@Chrismarsh

> The underlying issue is that some of the assumptions made in netcdf-cxx4 package.py result in system libnetcdf being linked to.

The assumption is that Spack compiler wrapper injects the required compiler and linker flags automatically and deprioritizes the system directories. If that doesn't work, there are problems in either your Spack configurations or in the core functionality of Spack. Obviously, I don't discard the third option that I miss something. The outputs of `readelf -d libnetcdf_c++4.so` before and after your changes could help figure that out.

> Specifically with migrating `hdf5/package.py` to cmake? or something else?

I was talking about `hdf5`, yes. Packages that support both `CMake` and `Autotools` are often not careful enough to deliver sufficiently similar results. The differences might affect the established workflows and use cases. The support for `CMake` in `netcdf-c` went much smoother as it left the choice to use `Autotools`.

> This includes removing the non-composable way mpicxx is used via direct manipulation of `CC=`.

Setting `CC`, `FC` and so on to the respective MPI compiler wrappers is a common practice. Note, however, that neither `mpi` nor `hdf5` is a direct dependency of `netcdf-cxx4` and, therefore, `libnetcdf_c++4.so` (or any other elf object that is installed) should not be linked to them and `netcdf-cxx4` should not `depends_on` them. This, for example, means that we should not set `CXX` to the MPI compiler wrapper (unless we want to support static-only installations like in https://github.com/spack/spack/pull/36485/commits/acdc5adc5223433cbbb2ae6ab9bf2e22f4d6e821 or #35466). The only reason for setting `CC` to the MPI compiler wrapper in `netcdf-cxx4` is to build a test program written in C, which is not installed and, therefore, the overlinking is not that much of an issue.

File: var/spack/repos/builtin/packages/netcdf-cxx4/package.py
Comment (Chrismarsh) [reply var/spack/repos/builtin/packages/netcdf-cxx4/package.py]: @scheibelp I kept this due to the library name being different from the package name, which, afaik, requires a custom libs property.

@skosukhin I removed this as it was not clear to me that throwing an exception here is expected spack behaviour. Specifically, [the](https://spack.readthedocs.io/en/latest/packaging_guide.html#custom-attributes) [docs](https://spack-tutorial.readthedocs.io/en/latest/tutorial_advanced_packaging.html#providing-libraries-to-dependents) don't recommend it, and most spack packages don't do this:

```
$ ag -A60 "libs\(self\)\:" | grep -i "spack.error"
netcdf-fortran/package.py:112-        raise spack.error.NoLibrariesError(
python/package.py:1054-        raise spack.error.NoLibrariesError(
libszip/package.py:35-            raise spack.error.NoLibrariesError(
parallel-netcdf/package.py:101-        raise spack.error.NoLibrariesError(
libaec/package.py:48-            raise spack.error.NoLibrariesError(
eccodes/package.py:301-        raise spack.error.NoLibrariesError(
hdf/package.py:126-            raise spack.error.NoLibrariesError(
serialbox/package.py:133-        raise spack.error.NoLibrariesError(msg.format(self.spec.name, self.spec.prefix))
```
Indeed, the only that do are packages maintained by you. Thus, I was following the broader community practice of not including it. I can see value in the error, and ultimately I really don't care. I can add it back if this is something you feel very strongly about.

Comment (skosukhin) [reply var/spack/repos/builtin/packages/netcdf-cxx4/package.py]: The broader community practice is to rely on the [default handler](https://github.com/spack/spack/blob/de3d1e6c66896f90f090c298fa05c5aefb42bde8/lib/spack/spack/spec.py#L1109C5-L1109C26), which raises [`NoLibrariesError`](https://github.com/spack/spack/blob/de3d1e6c66896f90f090c298fa05c5aefb42bde8/lib/spack/spack/spec.py#L1160).

**Comment by Chrismarsh** [comment]
Hi @skosukhin 

>Note, however, that neither mpi nor hdf5 is a direct dependency of netcdf-cxx4 and, therefore, libnetcdf_c++4.so (or any other elf object that is installed) should not be linked to them and netcdf-cxx4 should not depends_on them. 

I agree that it shouldn't, but it does. The `netcdf-cxx4` CMakeLists.txt explicitly [looks for](https://github.com/Unidata/netcdf-cxx4/blob/master/CMakeLists.txt#L407) and then [links against](https://github.com/Unidata/netcdf-cxx4/blob/master/cxx4/CMakeLists.txt#L38) `hdf5`. 

I was uncertain exactly how #29246 was interacting with `hdf5+mpi`. Indeed, I'd run into a similar issue with `gdal+netcdf ^hdf5+mpi` requiring the mpi wrappers (although I struggled to reproduce the error in a clean env).  Including `+mpi` when `^hdf5+mpi` seemed safest. 

But if you are certain it is only for when tests are enabled, then I think that's easily amended via
`depends_on('mpi', when='+tests ^hdf5+mpi')` and avoid touching `CC=`.

**Comment by skosukhin** [comment]
Hi @Chrismarsh

> I agree that it shouldn't, but it does. The netcdf-cxx4 CMakeLists.txt explicitly [looks for](https://github.com/Unidata/netcdf-cxx4/blob/master/CMakeLists.txt#L407) and then [links against](https://github.com/Unidata/netcdf-cxx4/blob/master/cxx4/CMakeLists.txt#L38) hdf5.

Yes, the `CMake` implementation of the package's build system has its own defects.

> `depends_on('mpi', when='+tests ^hdf5+mpi')` and avoid touching `CC=`

A `CMakePackage` obviously does not need to touch `CC`. This is from the `Autotools` world. I don't remember whether Spack populates `CMAKE_PREFIX_PATH` with all dependencies or the direct ones only. In the former case, I'm not sure whether it makes sense to add the `depends_on` directive for `mpi`.

File: var/spack/repos/builtin/packages/netcdf-cxx4/package.py
Comment (Chrismarsh) [reply var/spack/repos/builtin/packages/netcdf-cxx4/package.py]: Fair. I've made the change

**Comment by skosukhin** [comment]
@Chrismarsh I only wanted to make you aware of the potential issues. I don't mind if you have it your way. In that case, please, add yourself as a maintainer of the package.

**Comment by Chrismarsh** [comment]
@skosukhin For sure, I appreciate it! I'm just looking to try to help make spack as robust as possible. I am, of course, happy to help maintain if needed.

There is an interaction on macos with a homebrew (and I'm sure linux too) 'system' installed libnetcdf + libhdf5 that can be found via `*-config.cmake` files. If the autotools `netcdf-c` is built via spack, then the `*-config.cmake` files are not produced. Thus, cmake will use the homebrew libnetcdf-c/hdf5 instead of the spack versions. I have pushed a change that ensures netcdf-c build_system=cmake` is used. This is then very robust on all the machines I've tested.

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
> I am, of course, happy to help maintain if needed.

FWIW you can add yourself to

```
maintainers("WardF")
```

in the `netcdf-cxx4` package.py file if you want to make that explicit (this means for example that when users submit PRs related to this package, that our github bot would ping you as a potential reviewer).

**Pull request merged**
